### ISSUE TITLE
Sending POST requests exceeding 10KB results in 400 Bad Request under WSGI

### ISSUE THREAD

**Issue by ilatypov** [pallets/werkzeug#1068]
I get a 400 Bad Request on submitting a form with POST when the content length exceeds 10KB.

<pre><code>
[Mon Jan 30 10:00:21.917872 2017] [mpm_winnt:notice] [pid 6764:tid 612] AH00455: Apache/2.4.25 (Win32) OpenSSL/1.0.2j mod_wsgi/4.5.13 Python/2.7 configured -- resuming normal operations
[Mon Jan 30 10:00:21.918872 2017] [mpm_winnt:notice] [pid 6764:tid 612] AH00456: Server built: Dec 16 2016 18:55:21
[Mon Jan 30 10:00:21.918872 2017] [core:notice] [pid 6764:tid 612] AH00094: Command line: 'D:\\Apache24\\bin\\httpd.exe -d D:/Apache24'
[Mon Jan 30 10:00:21.926872 2017] [mpm_winnt:notice] [pid 6764:tid 612] AH00418: Parent: Created child process 3476
[Mon Jan 30 10:00:23.072938 2017] [mpm_winnt:notice] [pid 3476:tid 552] AH00354: Child: Starting 64 worker threads.
[Mon Jan 30 10:00:28.581253 2017] [wsgi:error] [pid 3476:tid 1196] [client 10.100.110.249:63814] mod_wsgi (pid=3476): Exception occurred processing WSGI script 'C:/HOME/werkzeug/werkzeug/testapp.wsgi'., referer: https://example.net:6443/testapp
[Mon Jan 30 10:00:28.582253 2017] [wsgi:error] [pid 3476:tid 1196] [client 10.100.110.249:63814] Traceback (most recent call last):, referer: https://example.net:6443/testapp
[Mon Jan 30 10:00:28.582253 2017] [wsgi:error] [pid 3476:tid 1196] [client 10.100.110.249:63814]   File "C:\\HOME\\werkzeug\\werkzeug\\testapp.py", line 260, in test_app, referer: https://example.net:6443/testapp
[Mon Jan 30 10:00:28.582253 2017] [wsgi:error] [pid 3476:tid 1196] [client 10.100.110.249:63814]     response = Response(render_testapp(req), mimetype='text/html'), referer: https://example.net:6443/testapp
[Mon Jan 30 10:00:28.582253 2017] [wsgi:error] [pid 3476:tid 1196] [client 10.100.110.249:63814]   File "C:\\HOME\\werkzeug\\werkzeug\\testapp.py", line 190, in render_testapp, referer: https://example.net:6443/testapp
[Mon Jan 30 10:00:28.583253 2017] [wsgi:error] [pid 3476:tid 1196] [client 10.100.110.249:63814]     req_form_info = escape('req.form: type %s, dir %s' % (type(req.form), dir(req.form))), referer: https://example.net:6443/testapp
[Mon Jan 30 10:00:28.583253 2017] [wsgi:error] [pid 3476:tid 1196] [client 10.100.110.249:63814]   File "C:\\HOME\\werkzeug\\werkzeug\\utils.py", line 73, in __get__, referer: https://example.net:6443/testapp
[Mon Jan 30 10:00:28.583253 2017] [wsgi:error] [pid 3476:tid 1196] [client 10.100.110.249:63814]     value = self.func(obj), referer: https://example.net:6443/testapp
[Mon Jan 30 10:00:28.583253 2017] [wsgi:error] [pid 3476:tid 1196] [client 10.100.110.249:63814]   File "C:\\HOME\\werkzeug\\werkzeug\\wrappers.py", line 523, in form, referer: https://example.net:6443/testapp
[Mon Jan 30 10:00:28.584253 2017] [wsgi:error] [pid 3476:tid 1196] [client 10.100.110.249:63814]     self._load_form_data(), referer: https://example.net:6443/testapp
[Mon Jan 30 10:00:28.584253 2017] [wsgi:error] [pid 3476:tid 1196] [client 10.100.110.249:63814]   File "C:\\HOME\\werkzeug\\werkzeug\\wrappers.py", line 371, in _load_form_data, referer: https://example.net:6443/testapp
[Mon Jan 30 10:00:28.584253 2017] [wsgi:error] [pid 3476:tid 1196] [client 10.100.110.249:63814]     mimetype, content_length, options), referer: https://example.net:6443/testapp
[Mon Jan 30 10:00:28.584253 2017] [wsgi:error] [pid 3476:tid 1196] [client 10.100.110.249:63814]   File "C:\\HOME\\werkzeug\\werkzeug\\formparser.py", line 195, in parse, referer: https://example.net:6443/testapp
[Mon Jan 30 10:00:28.585253 2017] [wsgi:error] [pid 3476:tid 1196] [client 10.100.110.249:63814]     content_length, options), referer: https://example.net:6443/testapp
[Mon Jan 30 10:00:28.585253 2017] [wsgi:error] [pid 3476:tid 1196] [client 10.100.110.249:63814]   File "C:\\HOME\\werkzeug\\werkzeug\\formparser.py", line 104, in wrapper, referer: https://example.net:6443/testapp
[Mon Jan 30 10:00:28.585253 2017] [wsgi:error] [pid 3476:tid 1196] [client 10.100.110.249:63814]     exhaust(), referer: https://example.net:6443/testapp
[Mon Jan 30 10:00:28.585253 2017] [wsgi:error] [pid 3476:tid 1196] [client 10.100.110.249:63814]   File "C:\\HOME\\werkzeug\\werkzeug\\wsgi.py", line 1123, in exhaust, referer: https://example.net:6443/testapp
[Mon Jan 30 10:00:28.585253 2017] [wsgi:error] [pid 3476:tid 1196] [client 10.100.110.249:63814]     self.read(chunk), referer: https://example.net:6443/testapp
[Mon Jan 30 10:00:28.585253 2017] [wsgi:error] [pid 3476:tid 1196] [client 10.100.110.249:63814]   File "C:\\HOME\\werkzeug\\werkzeug\\wsgi.py", line 1141, in read, referer: https://example.net:6443/testapp
[Mon Jan 30 10:00:28.586253 2017] [wsgi:error] [pid 3476:tid 1196] [client 10.100.110.249:63814]     return self.on_disconnect(), referer: https://example.net:6443/testapp
[Mon Jan 30 10:00:28.586253 2017] [wsgi:error] [pid 3476:tid 1196] [client 10.100.110.249:63814]   File "C:\\HOME\\werkzeug\\werkzeug\\wsgi.py", line 1109, in on_disconnect, referer: https://example.net:6443/testapp
[Mon Jan 30 10:00:28.586253 2017] [wsgi:error] [pid 3476:tid 1196] [client 10.100.110.249:63814]     raise ClientDisconnected(), referer: https://example.net:6443/testapp
[Mon Jan 30 10:00:28.586253 2017] [wsgi:error] [pid 3476:tid 1196] [client 10.100.110.249:63814] ClientDisconnected: 400 Bad Request: The browser (or proxy) sent a request that this server could not understand., referer: https://example.net:6443/testapp
</code></pre>

Including the sample test web application below.  It fails with an Internal Server Error when I click the Submit button.

<pre><code>
$ git diff HEAD werkzeug/testapp.py
diff --git a/werkzeug/testapp.py b/werkzeug/testapp.py
index 595555a..4503497 100644
--- a/werkzeug/testapp.py
+++ b/werkzeug/testapp.py
@@ -12,7 +12,6 @@
 import os
 import sys
 import werkzeug
-from textwrap import wrap
 from werkzeug.wrappers import BaseRequest as Request, BaseResponse as Response
 from werkzeug.utils import escape
 import base64
@@ -91,6 +90,21 @@ TEMPLATE = u'''\
   &lt;p>
     This page displays all available information about the WSGI server and
     the underlying Python interpreter.
+  &lt;h2 id"test-form">Test Form&lt;/h2>
+  &lt;p>
+  %(req_info)s
+  &lt;p>
+  %(req_data_info)s
+  &lt;p>
+  %(req_form_info)s
+  &lt;p>
+  &lt;table>%(wsgi_input)s&lt;/table>
+  &lt;p>
+  &lt;form method="POST">
+    &lt;input type="text" value="%(test_value)s" name="username"/>
+    Length: %(test_value_len)s
+    &lt;input type="submit" name="submit"/>
+  &lt;/form>
   &lt;h2 id="python-interpreter">Python Interpreter&lt;/h2>
   &lt;table>
     &lt;tr>
@@ -169,13 +183,29 @@ def render_testapp(req):
             escape(version)
         ))

+    req_info = escape('req: type %s, dir %s' % (type(req), dir(req)))
+    data = ''
+    # data = req.get_data()
+    req_data_info = escape('req.get_data(): type %s, dir %s' % (type(data), dir(data)))
+    req_form_info = escape('req.form: type %s, dir %s' % (type(req.form), dir(req.form)))
+    wsgi_input = []
+    sorted_input = sorted(req.form.items(),
+                            key=lambda x: repr(x[0]).lower())
+    for key, value in sorted_input:
+        wsgi_input.append('&lt;tr>&lt;th>%s&lt;td>&lt;code>%s&lt;/code>' % (
+            escape(str(key)),
+            ' '.join(escape(repr(value)))
+        ))
+
+    test_value = 'x' * (10241 - len('username=&submit=Submit+Query'));  # 10212
+
     wsgi_env = []
     sorted_environ = sorted(req.environ.items(),
                             key=lambda x: repr(x[0]).lower())
     for key, value in sorted_environ:
         wsgi_env.append('&lt;tr>&lt;th>%s&lt;td>&lt;code>%s&lt;/code>' % (
             escape(str(key)),
-            ' '.join(wrap(escape(repr(value))))
+            ' '.join(escape(repr(value)))
         ))

     sys_path = []
@@ -198,8 +228,14 @@ def render_testapp(req):
         'byteorder':        sys.byteorder,
         'werkzeug_version': werkzeug.__version__,
         'python_eggs':      '\n'.join(python_eggs),
+        'req_info':         req_info,
+        'req_data_info':    req_data_info,
+        'req_form_info':    req_form_info,
+        'test_value':       test_value,
+        'test_value_len':   len(test_value),
+        'wsgi_input':       '\n'.join(wsgi_input),
         'wsgi_env':         '\n'.join(wsgi_env),
-        'sys_path':         '\n'.join(sys_path)
+        'sys_path':         '\n'.join(sys_path),
     }).encode('utf-8')

$ git diff HEAD werkzeug/testapp.wsgi
diff --git a/werkzeug/testapp.wsgi b/werkzeug/testapp.wsgi
new file mode 100755
index 0000000..4faca39
--- /dev/null
+++ b/werkzeug/testapp.wsgi
@@ -0,0 +1,8 @@
+import sys, os
+path = os.path.abspath(os.path.join(os.path.dirname(__file__), ".."))
+if path not in sys.path:
+    print "Inserting path " + path
+    sys.path.insert(0, path)
+os.chdir(path)
+from werkzeug.testapp import test_app as application
+
</code></pre>

Including the Apache WSGI configuration below.

<pre><code>
Listen 6443 https
&lt;VirtualHost *:6443>
    WSGIScriptAlias /testapp C:/HOME/werkzeug/werkzeug/testapp.wsgi
    &lt;Directory C:/HOME/werkzeug/werkzeug>
        Require all granted
    &lt;/Directory>

    SSLEngine on
    ServerName example.net:6443
    ServerAlias example.net:6443
    SSLCertificateFile "${SRVROOT}/conf/ssl/cdc.crt"
    SSLCertificateKeyFile "${SRVROOT}/conf/ssl/cdc.key"
    DocumentRoot "${SRVROOT}/htdocs"
    CustomLog "${SRVROOT}/logs/ssl_request.log" \
          "%t %h %{SSL_PROTOCOL}x %{SSL_CIPHER}x \"%r\" %b"
&lt;/VirtualHost>
</code></pre>

As a work-around I figured I could place a req.get_data() call before accessing req.form.

**Comment by untitaker** [comment]
Have you tried this with any other servers? What makes you think this is a bug in Werkzeug?

**Comment by ilatypov** [comment]
Thanks, I submitted an issue with mod_wsgi.  In the meantime, tracking the (unexpeced) returned size worked for me (but I am not sure I covered all the cases).

<pre><code>
$ git diff werkzeug/wsgi.py
diff --git a/werkzeug/wsgi.py b/werkzeug/wsgi.py
index e64d64a..1c8f650 100644
--- a/werkzeug/wsgi.py
+++ b/werkzeug/wsgi.py
@@ -876,7 +876,9 @@ def _make_chunk_iter(stream, limit, buffer_size):
         stream = LimitedStream(stream, limit)
     _read = stream.read
     while 1:
+        print "_make_chunk_iter: buffer_size", buffer_size
         item = _read(buffer_size)
+        print "_make_chunk_iter: item len", len(item)
         if not item:
             break
         yield item
@@ -1077,6 +1079,7 @@ class LimitedStream(object):
     """

     def __init__(self, stream, limit):
+        print "LimitedStream stream:", str(type(stream))
         self._read = stream.read
         self._readline = stream.readline
         self._pos = 0
@@ -1133,13 +1136,20 @@ class LimitedStream(object):
         if size is None or size == -1:  # -1 is for consistence with file
             size = self.limit
         to_read = min(self.limit - self._pos, size)
+        read = None
         try:
+            print "to_read:", to_read
             read = self._read(to_read)
         except (IOError, ValueError):
             return self.on_disconnect()
-        if to_read and len(read) != to_read:
-            return self.on_disconnect()
+        finally:
+            if read is None:
+                print "self._read(to_read) threw an exception"
+            else:
+                print "len(read):", len(read)
         self._pos += len(read)
+        if to_read > 0 and len(read) == 0:
+            return self.on_disconnect()
         return read

     def readline(self, size=None):

</code></pre>

**Comment by untitaker** [comment]
Please do not file *bugs* on projects when you don't know where the bug is. Rather try to replace either mod_wsgi or werkzeug with an alternative (e.g. use a different server, or write a raw WSGI app without Werkzeug) and go from there.

**Comment by ilatypov** [comment]
Thanks again for narrowing down the root cause. I did find it odd that the code considers any size not matching the expected one to be an indication of stream exhaustion, but I guess I read too much into the code.  

"An empty string is returned when EOF is encountered immediately." https://docs.python.org/2/library/stdtypes.html#file.read

**Comment by hmoffatt** [comment]
What is the solution?

**Comment by davidism** [comment]
Looks like it was a mod_wsgi issue that was fixed a while ago. GrahamDumpleton/mod_wsgi#187 I  can't reproduce the original issue, uploading a 20 MB file with the following Flask app works as expected.

```python
from flask import Flask, redirect, request

app = Flask(__name__)

@app.route('/', methods=['GET', 'POST'])
def index():
    if request.method == 'POST':
        file = request.files['file']
        file.save('test.txt')
        return redirect('/')

    return '''
        <form method=post enctype=multipart/form-data>
            <input type=file name=file>
            <input type=submit>
        </form>
    '''
```

**Comment by hmoffatt** [comment]
Yes, I found that it was fixed in mod_wsgi 4.5.15, and my app is working again with long POSTs now that I've upgraded. Thanks. https://modwsgi.readthedocs.io/en/develop/release-notes/version-4.5.15.html#bugs-fixed
